File: src/deadline/client/job_bundle/loader.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

from __future__ import annotations

__all__ = [
    "read_yaml_or_json",
    "read_yaml_or_json_object",
    "parse_yaml_or_json_content",
    "validate_directory_symlink_containment",
]

import json
import os
from itertools import chain
from typing import Any, Dict, Optional, Tuple

import yaml

from ..exceptions import DeadlineOperationError


def validate_directory_symlink_containment(job_bundle_dir: str) -> None:
    """
    Validates the integrity of the job bundle, validating that all files
    contained in the job bundle resolve inside of the job bundle directory.
    """
    # The job bundle could itself be a symlink dir
    resolved_root = os.path.realpath(job_bundle_dir)
    if not os.path.isdir(resolved_root):
        raise DeadlineOperationError(
            f"Job bundle path provided is not a directory:\n{job_bundle_dir}"
        )
    for root_dir, dir_names, file_names in os.walk(resolved_root):
        for path in chain(dir_names, file_names):
            norm_path = os.path.normpath(os.path.join(root_dir, path))
            resolved_path = os.path.realpath(norm_path)
            common_path = os.path.commonpath([resolved_root, resolved_path])
            if common_path != resolved_root:
                raise DeadlineOperationError(
                    f"Job bundle cannot contain a path that resolves outside of the resolved bundle directory:\n{resolved_root}\n\nPath in bundle:\n{norm_path}\nResolves to:\n{resolved_path}"
                )


def read_yaml_or_json(job_bundle_dir: str, filename: str, required: bool) -> Tuple[str, str]:
    """
    Checks whether {filename}.json or {filename}.yaml exist in the provided
    job bundle directory, and returns a tuple (<file contents>, "YAML|JSON").

    Args:
        job_bundle_dir (str): The directory containing the job bundle.
        filename (str): The filename, without extension, to look for.
        required (bool): Whether this file is required. If not required and its missing,
                         the function returns ("", "").
    """
    path_prefix = os.path.join(job_bundle_dir, filename)
    has_json = os.path.isfile(path_prefix + ".json")
    has_yaml = os.path.isfile(path_prefix + ".yaml")
    if has_json and has_yaml:
        raise DeadlineOperationError(
            f"Job bundle directory has both {filename}.json and {filename}.yaml, only one is permitted:\n{job_bundle_dir}"
        )
    elif has_json:
        with open(path_prefix + ".json", encoding="utf8") as f:
            return (f.read(), "JSON")
    elif has_yaml:
        with open(path_prefix + ".yaml", encoding="utf8") as f:
            return (f.read(), "YAML")
    else:
        if required:
            raise DeadlineOperationError(
                f"Job bundle directory lacks a {filename}.json or {filename}.yaml:\n{job_bundle_dir}"
            )
        else:
            return ("", "")


def parse_yaml_or_json_content(file_contents: str, file_type: str, bundle_dir: str, filename: str):
    """
    Parses a yaml or json file that was read by the `read_yaml_or_json` function.

    Args:
        file_contents: The contents of the file that was read.
        file_type: The type of the file
        filename: The filename that was read, without extension (for error messages).
        bundle_dir: The bundle directory the file was in (for error messages).
    """
    if file_type == "JSON":
        try:
            return json.loads(file_contents)
        except json.JSONDecodeError as e:
            raise DeadlineOperationError(f"Error loading '{filename}.json':\n{e}")
    elif file_type == "YAML":
        try:
            return yaml.safe_load(file_contents)
        except yaml.MarkedYAMLError as e:
            raise DeadlineOperationError(f"Error loading '{filename}.yaml':\n{e}")
    else:
        raise RuntimeError(f"Unexpected file type '{file_type}' in job bundle:\n{bundle_dir}")


def read_yaml_or_json_object(
    bundle_dir: str, filename: str, required: bool
) -> Optional[Dict[str, Any]]:
    """
    Checks whether {filename}.json or {filename}.yaml exist in the provided
    job bundle directory, and returns the file parsed into an object.

    Args:
        job_bundle_dir (str): The directory containing the job bundle.
        filename (str): The filename, without extension, to look for.
        required (bool): Whether this file is required. If not required and its missing,
                         the function returns ("", "").
    """
    file_contents, file_type = read_yaml_or_json(bundle_dir, filename, required)
    if file_contents and file_type:
        return parse_yaml_or_json_content(file_contents, file_type, bundle_dir, filename)
    else:
        return None

File: crates/deadline-job-bundle/src/loader.rs
use deadline_api::errors::DeadlineError;
use std::fs;
use std::path::Path;

fn op_err(msg: String) -> DeadlineError {
    DeadlineError::OperationError(msg)
}

pub fn validate_directory_symlink_containment(job_bundle_dir: &Path) -> Result<(), DeadlineError> {
    let resolved_root = fs::canonicalize(job_bundle_dir).map_err(|_| {
        op_err(format!(
            "Job bundle path provided is not a directory:\n{}",
            job_bundle_dir.display()
        ))
    })?;
    if !resolved_root.is_dir() {
        return Err(op_err(format!(
            "Job bundle path provided is not a directory:\n{}",
            job_bundle_dir.display()
        )));
    }
    walk_and_check(&resolved_root, &resolved_root)
}

fn walk_and_check(resolved_root: &Path, dir: &Path) -> Result<(), DeadlineError> {
    let entries = fs::read_dir(dir)
        .map_err(|e| op_err(format!("Failed to read directory {}: {e}", dir.display())))?;
    for entry in entries {
        let entry = entry.map_err(|e| op_err(format!("Failed to read directory entry: {e}")))?;
        let path = entry.path();
        let resolved = fs::canonicalize(&path)
            .map_err(|e| op_err(format!("Failed to resolve path {}: {e}", path.display())))?;
        if !resolved.starts_with(resolved_root) {
            return Err(op_err(format!(
                "Job bundle cannot contain a path that resolves outside of the resolved bundle directory:\n{}\n\nPath in bundle:\n{}\nResolves to:\n{}",
                resolved_root.display(),
                path.display(),
                resolved.display()
            )));
        }
        if path.is_dir()
            && !path
                .symlink_metadata()
                .map(|m| m.is_symlink())
                .unwrap_or(false)
        {
            walk_and_check(resolved_root, &path)?;
        }
    }
    Ok(())
}

pub fn read_yaml_or_json(
    job_bundle_dir: &Path,
    filename: &str,
    required: bool,
) -> Result<(String, String), DeadlineError> {
    let base = job_bundle_dir.join(filename);
    let json_path = format!("{}.json", base.display());
    let yaml_path = format!("{}.yaml", base.display());
    let has_json = Path::new(&json_path).is_file();
    let has_yaml = Path::new(&yaml_path).is_file();

    match (has_json, has_yaml) {
        (true, true) => Err(op_err(format!(
            "Job bundle directory has both {filename}.json and {filename}.yaml, only one is permitted:\n{}",
            job_bundle_dir.display()
        ))),
        (true, false) => Ok((
            fs::read_to_string(&json_path)
                .map_err(|e| op_err(format!("Failed to read {json_path}: {e}")))?,
            "JSON".into(),
        )),
        (false, true) => Ok((
            fs::read_to_string(&yaml_path)
                .map_err(|e| op_err(format!("Failed to read {yaml_path}: {e}")))?,
            "YAML".into(),
        )),
        (false, false) if required => Err(op_err(format!(
            "Job bundle directory lacks a {filename}.json or {filename}.yaml:\n{}",
            job_bundle_dir.display()
        ))),
        _ => Ok((String::new(), String::new())),
    }
}

pub fn parse_yaml_or_json_content(
    file_contents: &str,
    file_type: &str,
    bundle_dir: &Path,
    filename: &str,
) -> Result<serde_json::Value, DeadlineError> {
    match file_type {
        "JSON" => serde_json::from_str(file_contents)
            .map_err(|e| op_err(format!("Error loading '{filename}.json':\n{e}"))),
        "YAML" => serde_yaml::from_str(file_contents)
            .map_err(|e| op_err(format!("Error loading '{filename}.yaml':\n{e}"))),
        _ => Err(op_err(format!(
            "Unexpected file type '{file_type}' in job bundle:\n{}",
            bundle_dir.display()
        ))),
    }
}

pub fn read_yaml_or_json_object(
    bundle_dir: &Path,
    filename: &str,
    required: bool,
) -> Result<Option<serde_json::Value>, DeadlineError> {
    let (contents, file_type) = read_yaml_or_json(bundle_dir, filename, required)?;
    if contents.is_empty() && file_type.is_empty() {
        Ok(None)
    } else {
        parse_yaml_or_json_content(&contents, &file_type, bundle_dir, filename).map(Some)
    }
}

pub fn save_yaml_or_json_to_file(
    bundle_dir: &Path,
    filename: &str,
    file_type: &str,
    data: &serde_json::Value,
) -> Result<(), DeadlineError> {
    let (ext, contents) = match file_type {
        "YAML" => ("yaml", deadline_yaml_dump(data)),
        "JSON" => (
            "json",
            serde_json::to_string_pretty(data)
                .map_err(|e| op_err(format!("Failed to serialize JSON: {e}")))?,
        ),
        _ => {
            return Err(op_err(format!(
                "Unexpected file type '{file_type}' in job bundle:\n{}",
                bundle_dir.display()
            )));
        }
    };
    let path = bundle_dir.join(format!("{filename}.{ext}"));
    fs::write(&path, &contents)
        .map_err(|e| op_err(format!("Failed to write {}: {e}", path.display())))
}

pub fn deadline_yaml_dump(data: &serde_json::Value) -> String {
    serde_yaml::to_string(data).unwrap_or_default()
}
